### PR TITLE
Add region event subscriptions

### PR DESCRIPTION
**Pull request by rbx**
Adds region event (creation, destruction) subscriptions interface and shared memory implementation.

- [x] Subscriptions via `void FairMQTransportFactory::SubscribeToRegionEvents(...)`.
- [x] Late subscribers get notified for all created regions.
- [x] Creation events.
- [x] Destruction events.
- [x] Custom user flags (`int64_t`) for `UnmanagedRegion`.
- [x] `FairMQTransportFactory::GetRegionInfo()` to get currently available regions.
- [x] Remove static state from shared memory manager, sockets, massages.

### CODE REVIEW

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- examples/region/Sampler.cxx  1
- fairmq/FairMQChannel.h  1
- fairmq/FairMQDevice.h  1
- examples/region/Sink.cxx  1
         

Clones removed
==============
+ fairmq/shmem/Socket.h  -2
+ fairmq/shmem/Socket.cxx  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357&bid=17788935)

File: fairmq/shmem/Manager.cxx [hunk 1]
```diff
@@ -94,84 +101,155 @@ void Manager::StartMonitor(const std::string& id)
     }
 }
 
-void Manager::Interrupt()
+pair<bipc::mapped_region*, uint64_t> Manager::CreateRegion(const size_t size, const int64_t userFlags, RegionCallback callback, const string& path /* = "" */, int flags /* = 0 */)
 {
-}
+    try {
+        uint64_t id = 0;
```
Comment (dennisklein) [inline fairmq/shmem/Manager.cxx:107]: Codacy Issue found: [The scope of the variable 'id' can be reduced.](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- examples/region/Sampler.cxx  1
- fairmq/FairMQDevice.h  1
- fairmq/FairMQChannel.h  1
- examples/region/Sink.cxx  1
         

Clones removed
==============
+ fairmq/shmem/Socket.cxx  -1
+ fairmq/shmem/Socket.h  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357&bid=17788935)

File: fairmq/shmem/Manager.cxx [hunk 2]
```diff
@@ -94,84 +101,155 @@ void Manager::StartMonitor(const std::string& id)
...
+
+            if (rc) {
+                LOG(debug) << "region counter found, with value of " << rc->fCount << ". incrementing.";
+                (rc->fCount)++;
+                LOG(debug) << "incremented region counter, now: " << rc->fCount;
+            } else {
+                LOG(debug) << "no region counter found, creating one and initializing with 1";
+                rc = fManagementSegment.construct<RegionCounter>(bipc::unique_instance)(1);
+                LOG(debug) << "initialized region counter with: " << rc->fCount;
+            }
```
Comment (dennisklein) [inline fairmq/shmem/Manager.cxx:124]: This looks like a race condition, does it not? What if another process constructs the `RegionCounter` between the find and the construct?

File: fairmq/shmem/Manager.cxx [hunk 3]
```diff
@@ -94,84 +101,155 @@ void Manager::StartMonitor(const std::string& id)
...
-        return &(r.first->second->fRegion);
+    } catch (bipc::interprocess_exception& e) {
+        LOG(error) << "cannot create region. Already created/not cleaned up?";
+        LOG(error) << e.what();
+        throw;
     }
 }
 
-Region* Manager::GetRemoteRegion(const uint64_t id)
+void Manager::RemoveRegion(const uint64_t id)
```
Comment (dennisklein) [inline fairmq/shmem/Manager.cxx:155]:
```suggestion
void Manager::RemoveRegionUnsafe(const uint64_t id)
```
?

File: fairmq/shmem/Manager.cxx (hunk 2 above)
Comment (rbx) [reply fairmq/shmem/Manager.cxx:124]: There is a mutex a couple of lines above.

Comment (rbx) [reply fairmq/shmem/Manager.cxx:124]: line 112.

File: fairmq/shmem/Manager.cxx (hunk 3 above)
Comment (rbx) [reply fairmq/shmem/Manager.cxx:155]: Added a lock here also. For now no need for the unsafe method.

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/FairMQDevice.h  1
- examples/region/Sink.cxx  1
- examples/region/Sampler.cxx  1
- fairmq/FairMQChannel.h  1
         

Clones removed
==============
+ fairmq/shmem/Socket.cxx  -1
+ fairmq/shmem/Socket.h  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357&bid=17788935)

File: fairmq/shmem/Manager.cxx (hunk 2 above)
Comment (dennisklein) [reply fairmq/shmem/Manager.cxx:124]: kk

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- fairmq/FairMQChannel.h  1
- examples/region/Sink.cxx  1
- fairmq/FairMQDevice.h  1
- examples/region/Sampler.cxx  1
         

Clones removed
==============
+ fairmq/shmem/Socket.cxx  -1
+ fairmq/shmem/Socket.h  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357&bid=17788935)

**Comment by dennisklein** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- examples/region/Sink.cxx  1
- examples/region/Sampler.cxx  1
- fairmq/FairMQDevice.h  1
- fairmq/FairMQChannel.h  1
         

Clones removed
==============
+ fairmq/shmem/Socket.h  -2
+ fairmq/shmem/Socket.cxx  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/FairRootGroup/FairMQ/pullRequest?prid=5424357&bid=17788935)

**Pull request merged**